### ISSUE TITLE
Prebuilt libraries for BRN

### ISSUE THREAD

**Issue by CedricGuillemet** [BabylonJS/BabylonNative#1270]
In order to speed up BRN builds, I want to evaluate the feasibility of having (some) prebuilt libraries of BabylonNative.
As discussed with @bghgary , because of JS engine, it won't be possible to prebuild everything. Does it make sense to have more clear separation for some libraries (JS engine layer + internal functionnality).

List of libraries dependent of JSEngine:

- NativeXR
- NativeCamera
- NativeEngine
- napi
- Console
- JSRuntime
- NativeOptimizations
- v8inspector
- ScriptLoader
- XMLHttpRequest 
- AppRuntime
- Graphics
- Window
- Canvas
- AbortController
- Externaltexture
- NativeTracing

Independent ones:

- edtaa3
- etc1
- tinyexr
- pvrtc
- OGLCompiler
- OSDependent
- GenericCodeGen
- llhttp(?)
- squish
- iqa
- arcana
- bx
- etc2
- astc-encoder
- nvtt
- spirv-cross-glsl
- urllib
- bimg
- MachineIndependent
- glslang
- openxr_loader
- bgfx
- SPIRV
- xr
- URL(?)
- WebSocket(?)

**Comment by ryantrem** [comment]
Would this specifically be local builds, build agents, or both?

**Comment by ryantrem** [comment]
If CI VMs are the main concern, could we use GitHub workflow caching for the build output? This would effectively make the builds incremental. I’m not sure how hard it wold be to generate a correct cache key for build outputs though. https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#

**Comment by CedricGuillemet** [comment]
Yes, it makes sense to save some binaries when a PR is merged, using BN commit hash as a key. And reuse with all builds following as long as it's the same BN commit. BN is not updated very often so it makes complete sense.

**Comment by CedricGuillemet** [comment]
I've used MSVC to get local build times. I'm keeping here libraries with >2seconds times.

squish 6
iqa 6
llhttp 8.7
spirv-cross-core 12
arcana 10
bx 13
astc-encoder 24
nvtt 11.7
spirv-cross-glsl 16.4
url-lib 18.8
spirv-cross-hlsl 7
bimg 11
openxr_loader 40
MachineIndependent 37
bgfx 11.3
xr 9
SPIRV 11.4

253.3 seconds





v8inspector 25
XMLHttpRequest 6.5
ScriptLoader 6.8
AppRuntime 6.6
Graphics 13.4
NativeXr 6.2
NativeCamera 8.2
Canvas 13.6
NativeEngine 19.4
native-input 10.5

116.2 seconds

so ~70% of build time can be cached and reused

**Comment by CedricGuillemet** [comment]
We discussed that with @bghgary and because of changes in JS interface, react,... this might not be possible.

**Comment by CedricGuillemet** [comment]
I did some tests today with Github Caching and got these issues:
- Total cache per repo is limited to 10Gb
- cache untouched for 7 days will be removed
- CMake does changes outside the build directory that can't be cached (node install for example)

**Comment by CedricGuillemet** [comment]
I tried to reuse, locally, bgfx static libraries.
First, I build the PG as usual, then, I change the cmake file in dependencies to reuse the library and headers.
```
set(BGFX_DIR "${CMAKE_CURRENT_SOURCE_DIR}/bgfx.cmake/bgfx" CACHE STRING "Location of bgfx." )
set(BIMG_DIR "${CMAKE_CURRENT_SOURCE_DIR}/bgfx.cmake/bimg" CACHE STRING "Location of bimg." )
set(BX_DIR "${CMAKE_CURRENT_SOURCE_DIR}/bgfx.cmake/bx" CACHE STRING "Location of bx." )

add_library(bgfx STATIC IMPORTED)
set_property(TARGET bgfx PROPERTY IMPORTED_LOCATION "${CMAKE_CURRENT_BINARY_DIR}/Dependencies/bgfx.cmake/Debug/bgfx.a")
target_include_directories(bgfx INTERFACE "${BGFX_DIR}/include")
```
But it didn't work and include is not added to the build command line
and I end up seeing this error:
```
BgfxCallback.h:6:10: fatal error: 'bgfx/bgfx.h' file not found
#include <bgfx/bgfx.h>
```
Any idea why it happens ?

**Comment by CedricGuillemet** [comment]
WIP using prebuilt on win32. bgfx, glslang and spirv-cross
before:
![image](https://github.com/BabylonJS/BabylonNative/assets/1312968/c2954f70-185d-41d3-9cdd-ce503c77a78c)
after:
![image](https://github.com/BabylonJS/BabylonNative/assets/1312968/e9ceb4ee-c44e-44eb-8211-155212fa7cdf)

**Comment by CedricGuillemet** [comment]
I'm reaching a point where cmake needs to run to fetch content (jsruntimehost, arcana,...) while also needing the prebuilt cache so it can import the libraries. But the cache (in order to be invalidated) needs a commit hash. A way to do it is to run cmake 2 times:
- first to fetch content
- second to import library or create build projects for 3rd party libs.
That's not something I'd like to do because of less maintainable cmake scripts and the benefits will be low (around 3 minutes per build)
The problem is more BRN builds duration than BN ones.
